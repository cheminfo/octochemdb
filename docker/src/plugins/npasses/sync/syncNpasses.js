import md5 from 'md5';

import debugLibrary from '../../../utils/Debug.js';
import createIndexes from '../../../utils/createIndexes.js';
import { shouldUpdate } from '../../../utils/shouldUpdate.js';
import { getTaxonomiesForCmaupsAndNpasses } from '../../activesOrNaturals/utils/utilsTaxonomies/getTaxonomiesForCmaupsAndNpasses.js';
import { taxonomySynonyms } from '../../activesOrNaturals/utils/utilsTaxonomies/taxonomySynonyms.js';
import { getNormalizedActivities } from '../../cmaups/sync/utils/getNormalizedActivities.js';

import npassesStartSync from './utils/npassesStartSync.js';
import { parseNpasses } from './utils/parseNpasses.js';
/**
 * @description Synchronizes the npasses collection with the NPASS database
 * @param {*} connection - mongo connection
 * @returns {Promise} returns npasses collection
 */
export async function sync(connection) {
  const debug = debugLibrary('syncNpasses');
  try {
    const {
      lastDocumentImported,
      progress,
      sources,
      collection,
      general,
      activities,
      properties,
      speciesPair,
      speciesInfo,
      targetInfo,
      logs,
    } = await npassesStartSync(connection);
    let isTimeToUpdate = await shouldUpdate(
      progress,
      sources,
      lastDocumentImported,
      process.env.NPASS_UPDATE_INTERVAL,
      connection,
    );
    let counter = 0;
    let imported = 0;
    let start = Date.now();
    // Reimport collection again only if lastDocument imported changed or importation was not completed
    if (isTimeToUpdate) {
      const oldToNewTaxIDs = await taxonomySynonyms();
      const collectionTaxonomies = await connection.getCollection('taxonomies');
      // create temporary collection
      const temporaryCollection = await connection.getCollection('npasses_tmp');
      debug.info(`Start parsing npasses`);
      // set progress to updating
      progress.state = 'updating';
      await connection.setProgress(progress);
      // parse npasses
      for await (const entry of parseNpasses(
        general,
        activities,
        properties,
        speciesPair,
        speciesInfo,
        targetInfo,
        connection,
      )) {
        counter++;
        // if test mode, stop after 20 entries
        if (process.env.NODE_ENV === 'test' && counter > 20) break;

        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          debug.trace(
            `Processing: counter: ${counter} - imported: ${imported}`,
          );
          start = Date.now();
        }
        /// Normalize Taxonomies
        if (entry.data.taxonomies) {
          let taxonomies = await getTaxonomiesForCmaupsAndNpasses(
            entry,
            collectionTaxonomies,
            oldToNewTaxIDs,
            'npasses',
          );
          entry.data.taxonomies = taxonomies;
        }
        // Normalize activities
        if (entry.data.activities) {
          let activities = await getNormalizedActivities(
            entry,
            collectionTaxonomies,
            oldToNewTaxIDs,
          );
          entry.data.activities = activities;
        }
        entry._seq = ++progress.seq;
        // import entry to temporary collection
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );

        imported++;
      }
      // rename temporary collection to npasses
      await temporaryCollection.rename('npasses', {
        dropTarget: true,
      });
      // set progress to updated
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      await connection.setProgress(progress);
      // set logs
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      // create indexes
      await createIndexes(collection, [
        { 'data.ocl.noStereoTautomerID': 1 },
        { _seq: 1 },
      ]);

      debug.info(`npasses collection updated`);
    } else {
      debug.info(`file already processed`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'npasses',
        connection,
        stack: e.stack,
      });
    }
  }
}
