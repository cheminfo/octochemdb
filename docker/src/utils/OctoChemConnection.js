import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
