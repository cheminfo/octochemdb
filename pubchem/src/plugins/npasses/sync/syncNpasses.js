import Debug from '../../../utils/Debug.js';
import npassesStartSync from './utils/npassesStartSync.js';

import { parseNpasses } from './utils/parseNpasses.js';

const debug = Debug('syncNpasses');

export async function sync(connection) {
  try {
    const {
      firstID,
      lastDocumentImported,
      progress,
      sources,
      collection,
      general,
      activities,
      properties,
      speciesPair,
      speciesInfo,
      logs,
    } = await npassesStartSync(connection);
    await collection.createIndex({ ' _seq': 1 });

    // we reparse all the file and skip if required
    let skipping = firstID !== undefined;
    let counter = 0;
    let imported = 0;
    let start = Date.now();

    // Reimport collection again only if lastDocument imported changed or importation was not completed
    if (
      lastDocumentImported === null ||
      sources !== progress.sources ||
      progress.state !== 'updated'
    ) {
      let parseSkip;
      if (skipping && progress.state !== 'updated') {
        parseSkip = firstID;
      }
      debug(`Start parsing npass`);

      for await (const entry of parseNpasses(
        general,
        activities,
        properties,
        speciesPair,
        speciesInfo,
        parseSkip,
        connection,
      )) {
        counter++;
        if (process.env.TEST === 'true' && counter > 20) break;

        if (
          Date.now() - start >
          Number(process.env.DEBUG_THROTTLING || 10000)
        ) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }

        entry._seq = ++progress.seq;
        progress.state = 'updating';
        await collection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );
        await connection.setProgress(progress);
        imported++;
      }
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      progress.sources = sources;
      progress.date = new Date();
      progress.state = 'updated';
      await connection.setProgress(progress);
      debug(`${imported} compounds processed`);
    } else {
      debug(`file already processed`);
    }
    // we remove all the entries that are not imported by the last file
    const result = await collection.deleteMany({
      _seq: { $lte: logs.startSequenceID },
    });
    debug(`Deleting entries with wrong source: ${result.deletedCount}`);
  } catch (e) {
    const optionsDebug = { collection: 'npasses', connection };
    debug(e, optionsDebug);
  }
}
