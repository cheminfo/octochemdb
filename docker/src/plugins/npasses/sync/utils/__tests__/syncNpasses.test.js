import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../../utils/OctoChemConnection.js';
import { sync } from '../../syncNpasses';

test(
  'syncNpasses',
  async () => {
    const connection = new OctoChemConnection();
    const taxonomiesCollection = await connection.getCollection('taxonomies');
    while (true) {
      if ((await taxonomiesCollection.countDocuments()) === 20) {
        break;
      }
    }
    await sync(connection);
    const collection = await connection.getCollection('npasses');
    const collectionEntry = await collection
      .find({ _id: 'NPC100380' })
      .limit(1);
    const result = await collectionEntry.next();
    if (result?._seq) {
      delete result._seq;
      delete result.data.ocl.coordinates; // they change every time
    }
    expect(result).toMatchSnapshot();
    await connection.close();
  },
  { timeout: 30000 },
);
