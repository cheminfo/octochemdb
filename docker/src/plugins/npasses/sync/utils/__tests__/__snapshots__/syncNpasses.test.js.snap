// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncNpasses 1`] = `
{
  "_id": "NPC10005",
  "data": {
    "activities": [
      {
        "assay": "MIC : 100 ug/ml",
        "dbRef": {
          "$id": "NPC10005",
          "$ref": "npasses",
        },
        "externalRef": "PMID : 10650093",
      },
      {
        "assay": "MIC : 100 ug/ml",
        "dbRef": {
          "$id": "NPC10005",
          "$ref": "npasses",
        },
        "externalRef": "PMID : 10650093",
      },
    ],
    "cid": "13915595",
    "ocl": {
      "idCode": "e\`rpD@@@DLbbRbTbReTRTJbRdJVTrbRtUQcfURqtNHjkUUUUUUUUUUUUUUUTtHPiHxheUAhBlX@@",
      "noStereoTautomerID": "e\`rpD@@@JLdTTTbreTRTJbRdJVTrbRtURRadVPrwtKHmUUUUUUUUUUUUUUUTt@Aj\`QbcGzPYQTq|L_C@",
    },
    "taxonomies": [
      {
        "dbRef": {
          "$id": "NPC10005",
          "$ref": "npasses",
        },
        "family": "Chenopodiaceae",
        "genus": "Haloxylon",
        "kingdom": "Viridiplantae",
        "species": "Haloxylon salicornicum",
      },
    ],
  },
}
`;
